### PR TITLE
restore overlapping dkg + validators update bug fix

### PR DESCRIPTION
**Pull request by jinmannwong**

### CODE REVIEW

File: consensus/state.go [hunk 1]
```diff
@@ -1146,6 +1152,10 @@ func (cs *State) getEntropy(height int64) *types.ChannelEntropy {
 		if entropy.Height != height {
 			panic(fmt.Sprintf("Height mismatch found between entropy and the height of entropy requested! %v vs %v", entropy.Height, height))
 		}
+		if cs.nextAeonStartFunc != nil && entropy.Entropy.NextAeonStart == 0 {
+			entropy.Entropy.NextAeonStart = cs.nextAeonStartFunc(height)
+			cs.newEntropy[height] = entropy
+		}
```
Comment (jinmannwong) [inline consensus/state.go:1158]: Consensus state fills in NextAeonStart inside the BlockEntropy by querying the dkgRunner.

File: beacon/dkg_runner.go [hunk 2]
```diff
@@ -110,7 +110,7 @@ func (dkgRunner *DKGRunner) FastSync(blockStore sm.BlockStore) error {
 		return fmt.Errorf("FastSync: dkgRunner running!")
 	}
 
-	dkgHeight := dkgRunner.aeonEnd
+	dkgHeight := dkgRunner.aeonStart
```
Comment (jinmannwong) [inline beacon/dkg_runner.go:113]: Reverting changes that forced non-overlapping DKGs

File: beacon/dkg_runner.go [hunk 3]
```diff
@@ -172,6 +172,17 @@ func (dkgRunner *DKGRunner) OnBlock(blockHeight int64, entropy types.ThresholdSi
...
+func (dkgRunner *DKGRunner) NextAeonStart(height int64) int64 {
+	dkgRunner.mtx.Lock()
+	defer dkgRunner.mtx.Unlock()
+
+	if height != dkgRunner.height+1 {
+		panic(fmt.Sprintf("consensus state requested next aeon start for unexpected height %v. Expected %v", height, dkgRunner.height+1))
+	}
+	return dkgRunner.aeonStart
+}
+
```
Comment (jinmannwong) [inline beacon/dkg_runner.go:185]: Function for passing the NextAeonStart to consensus state for inclusion into BlockHeader. This allows app to trigger validator changeovers required before the start of the aeon.

File: consensus/replay.go [hunk 4]
```diff
@@ -329,6 +329,7 @@ func (h *Handshaker) ReplayBlocks(
 				}
 				state.Validators = types.NewValidatorSet(vals)
 				state.NextValidators = types.NewValidatorSet(vals)
+				state.DKGValidators = types.NewValidatorSet(vals)
```
Comment (jinmannwong) [inline consensus/replay.go:332]: Stefanos' bug fix

File: beacon/slot_protocol_enforcer.go [hunk 5]
```diff
@@ -29,20 +29,20 @@ const (
...
 }
 
-func NewSlotProtocolEnforcer() *SlotProtocolEnforcer{
+func NewSlotProtocolEnforcer() *SlotProtocolEnforcer {
 	return &SlotProtocolEnforcer{
-	alreadySeenMsgs     : make(map[string]struct{}),
-	pendingTxs          : make([]*pendingTx, 0),
-	cbWhenUpdated       : nil,
-	logger              : log.NewNopLogger(),
+		alreadySeenMsgs: make(map[string]struct{}),
```
Comment (n-hutton) [inline beacon/slot_protocol_enforcer.go:42]: is this from running make fmt?

Comment (jinmannwong) [reply beacon/slot_protocol_enforcer.go:42]: Yes

**Pull request merged**